flixel/system/scaleModes/ShaderScaleMode.hx

```diff
@@ -0,0 +1,296 @@
+	{
+		filters.push(this.filter);
+	}
+
+	public function deactivate():Void
+	{
+		filters.remove(this.filter);
+	}
+
+	override private function updateScaleOffset():Void
+	{
+		scale.x = 1;
+		scale.y = 1;
+	}
+
+	override private function updateGameSize(Width:Int, Height:Int):Void 
+	{
+		gameSize.x = FlxG.width;
+		gameSize.y = FlxG.height;
+	}
+
+	override private function updateOffsetY():Void
+	{
+		var tempOffsetY = 0.0;
+    	for (i in 0...FlxG.game.numChildren)
+    	{
+        	var child = FlxG.game.getChildAt(i);
+
+        	if(child.y < 0 && child.y < tempOffsetY)
+        	{
+        		tempOffsetY = child.y;
+        	}
+    	}
+
+    	pointerOffset.y = tempOffsetY * (scaleY - 1) / scaleY;
+		offset.y = tempOffsetY * (scaleY - 1);
+	}
+
+	override private function updateOffsetX():Void
+	{
+		pointerOffset.x = 0;
+		offset.x = 0;
+	}
+	#else
+	private var scaleX:Float;
+	private var scaleY:Float;
+	private var strength:Float;
+
+
+	private var postProcess:PostProcess;
+
+	private static inline var VERTEX_SHADER:String = "
+#ifdef GL_ES
+	precision mediump float;
+#endif
+
+attribute vec2 aVertex;
+attribute vec2 aTexCoord;
+varying vec2 vTexCoord;
+
+void main() {
+	vTexCoord = aTexCoord;
+	gl_Position = vec4(aVertex, 0.0, 1.0);
+}";
+	private static inline var NEAREST_SHADER:String = "varying vec2 vTexCoord;
+//declare uniforms
+uniform sampler2D uImage0;
+uniform vec2 uResolution;
+
+uniform float scaleX;
+uniform float scaleY;
+uniform float strength;
+
+void main()
+{
+    vec2 tc = vec2(vTexCoord.x,vTexCoord.y+scaleY-1.0);
+    
+    gl_FragColor = texture2D(uImage0, vec2(tc.x/scaleX*strength,tc.y/scaleY*strength));
+}";
+	private static inline var BILINEAR_SHADER:String = "varying vec2 vTexCoord;
+//declare uniforms
+uniform sampler2D uImage0;
+uniform vec2 uResolution;
+
+uniform float scaleX;
+uniform float scaleY;
+uniform float strength;
+
+void main()
+{
+    vec2 tc = vec2(vTexCoord.x,vTexCoord.y+scaleY-1.0);
+
+
+	float OneTexelX = 1.0/uResolution.x*strength;
+	float OneTexelY = 1.0/uResolution.y*strength;
+
+	vec2 coord1 = vec2(tc.x/scaleX,tc.y/scaleY)+vec2(0.0, OneTexelY);
+	vec2 coord2 = vec2(tc.x/scaleX,tc.y/scaleY)+vec2(OneTexelX, 0.0 );
+	vec2 coord3 = vec2(tc.x/scaleX,tc.y/scaleY)+vec2(OneTexelX, OneTexelY);
+	vec2 coord4 = vec2(tc.x/scaleX,tc.y/scaleY);
+	
+	vec4 s1 = vec4(texture2D(uImage0, coord1));
+	vec4 s2 = vec4(texture2D(uImage0, coord2));
+	vec4 s3 = vec4(texture2D(uImage0, coord3));
+	vec4 s4 = vec4(texture2D(uImage0, coord4));
+	
+	vec2 Dimensions = vec2(tc) * uResolution;
+	
+	float fu = fract(Dimensions.x);
+	float fv = fract(Dimensions.y);
+	
+	vec4 tmp1 = mix(s4, s2, fu);
+	vec4 tmp2 = mix(s1, s3, fu);
+	
+	vec4 t0 = mix(tmp1, tmp2, fv);
+    
+    gl_FragColor = t0;//texture2D(uImage0, vec2(tc.x/scaleX,tc.y/scaleY));
+}";
+
+	public function new(shader:OneOfTwo<ShaderScaleEnum, Shader> = ShaderScaleEnum.NEAREST, scaleX:Float = 1, scaleY:Float = 1)
+	{
+		super();
+
+		if (Std.is(shader, Shader))
```

Comment:
Isn't this unnecessary since we're leveraging Haxe's type system to ensure people send valid Shader types?